google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java

```diff
@@ -0,0 +1,3482 @@
+            .build();
+      default:
+        throw SpannerExceptionFactory.newSpannerException(
+            ErrorCode.INVALID_ARGUMENT, "Unsupported cloud type: " + cloudTypeProto);
+    }
+  }
+
+  /** Unmarshall ByteString to serializable object. */
+  private <T extends Serializable> T unmarshall(ByteString input)
+      throws IOException, ClassNotFoundException {
+    ObjectInputStream objectInputStream = new ObjectInputStream(input.newInput());
+    return (T) objectInputStream.readObject();
+  }
+
+  /** Marshall a serializable object into ByteString. */
+  private <T extends Serializable> ByteString marshall(T object) throws IOException {
+    ByteString.Output output = ByteString.newOutput();
+    ObjectOutputStream objectOutputStream = new ObjectOutputStream(output);
+    objectOutputStream.writeObject(object);
+    objectOutputStream.flush();
+    objectOutputStream.close();
+    return output.toByteString();
+  }
+
+  /** Build Timestamp from micros. */
+  private Timestamp timestampFromMicros(long micros) {
+    long seconds = TimeUnit.MICROSECONDS.toSeconds(micros);
+    int nanos = (int) (micros * 1000 - seconds * 1000000000);
+    return Timestamp.ofTimeSecondsAndNanos(seconds, nanos);
+  }
+
+  /** Build TimestampBound from Concurrency. */
+  private TimestampBound timestampBoundsFromConcurrency(Concurrency concurrency) {
+    if (concurrency.hasStalenessSeconds()) {
+      return TimestampBound.ofExactStaleness(
+          (long) (concurrency.getStalenessSeconds() * 1000000), TimeUnit.MICROSECONDS);
+    } else if (concurrency.hasMinReadTimestampMicros()) {
+      return TimestampBound.ofMinReadTimestamp(
+          timestampFromMicros(concurrency.getMinReadTimestampMicros()));
+    } else if (concurrency.hasMaxStalenessSeconds()) {
+      return TimestampBound.ofMaxStaleness(
+          (long) (concurrency.getMaxStalenessSeconds() * 1000000), TimeUnit.MICROSECONDS);
+    } else if (concurrency.hasExactTimestampMicros()) {
+      return TimestampBound.ofReadTimestamp(
+          timestampFromMicros(concurrency.getExactTimestampMicros()));
+    } else if (concurrency.hasStrong()) {
+      return TimestampBound.strong();
+    } else if (concurrency.hasBatch()) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INVALID_ARGUMENT,
+          "batch mode should not be in snapshot transaction: " + concurrency);
+    }
+    throw SpannerExceptionFactory.newSpannerException(
+        ErrorCode.INVALID_ARGUMENT, "Unsupported concurrency mode: " + concurrency);
+  }
+
+  /** Get DatabaseId by parsing the given dbPath. */
+  private static DatabaseId getDatabaseId(String dbPath) {
+    Matcher matcher = DB_NAME.matcher(dbPath);
+    if (!matcher.matches()) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INVALID_ARGUMENT, "Bad database path: " + dbPath);
+    }
+    return DatabaseId.of(matcher.group(1), matcher.group(2), matcher.group(3));
```

Comment:
Can we directly use https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseId.java#L84?